File: 70-climbing-stairs/go/main.go
package main

import "fmt"

var cache = map[int]int{}

func main() {
	fmt.Println(climbStairs(5))
}

func climbStairs(n int) int {
	if n == 1 {
		return 1
	}

	if n == 2 {
		return 2
	}

	if val, ok := cache[n]; ok {
		return val
	}

	cache[n] = climbStairs(n-1) + climbStairs(n-2)
	return cache[n]
}

File: 70-climbing-stairs/python/main.py
def main():
    solution = Solution()
    print(solution.climbStairs(5))
    print(solution.climbStairs(3))

class Solution:
    cache = {}

    def climbStairs(self, n: int) -> int:
        if n <= 2:
            return n

        if n in self.cache.keys():
            return self.cache[n]

        self.cache[n] = self.climbStairs(n-1) + self.climbStairs(n-2)
        return self.cache[n]



# Different implemnetioan
class Solution2:
    def climbStairs(self, n: int) -> int:
        current, previous = 1, 0

        for i in range(n):
            previous, current = current, previous+current
        return current

main()